Add Scene_Settings - Config 3 of N

Ghabry opened the pull request:
Fix #666
Fix #713
Fix #794
Fix #1134
Fix #1891
Fix #2721
Fix #2843
Fix #2844
Fix #2883

----------

This is a WIP to add a settings menu. Early feedback is welcome on the general layout and feature set this should have.

Input:
- [x] Add input remapping for keyboard
- [x] Add input remapping for gamepads

Audio:
- [x] BGM Volume
- [x] SFX Volume

_Not planned_:
- [ ] Configure Midi backend to use
- [ ] Configure Midi soundfont to use
- [ ] Configure which system audio device to output to.
- [ ] Number of output channels
- [ ] Output sample rate
- [ ] Output bit depth
- [ ] Bit-perfect output toggle?

Video:
- [x] Show which renderer is active and leave future support for multiple renderers
- [x] Fullscreen
- [ ] Screen Resolution
- [x] Vsync
- [x] Frame Limiter
- [x] Show FPS

Access:
- [x] Game Browser
- [x] Title Screen
- [x] Menu Scene
- [x] OpenVideoOption event command
- [x] Add a key to open the settings menu

Persistence:
- [x] Load settings to user config file
- [x] Save settings

Others:
- [ ] Refresh UI if video options changed by key press
- [x] Show title screen in background

lychees commented:
FYI, there is a setting menu in the yumebooru's multiplayer mode which already include some features (i.e. SFX Volume).
https://github.com/CataractJustice/ynoclient/blob/master/src/game_multiplayer_settings_scene.cpp

Ghabry commented:
Jenkins: Test this please <3

fdelapena commented:
So far so good :+1: . In games without title scene, is it showing the default pixel background color? What about games skipping title scene?

Fullscreen does not toggle on Emscripten yet, likely expected (disabling it on emscripten for now might be worth).

Regarding the checklist, output channels is relative to single sample output channels? (16 unique samples, 4 exclusive same-sample channel, total up to 64)  or midi poliphony? (fluidsynth default is 256) or just global output channels meaning mono (1) / stereo (2) / surround (n)?.

Additional ideas, for this or for later (config 4 of N), e.g. toggle pause on focus lost and other:
Click here: https://github.com/EasyRPG/Player/labels/Settings

jetrotal commented:
Hey guys, silly suggestion:
what about making the look of it like this:
![image](https://user-images.githubusercontent.com/45118493/157803522-e068a0c9-d5e3-4fc1-990a-561646bb0d30.png)

It's a mockup based on the game select window:
![image](https://user-images.githubusercontent.com/45118493/157803622-85dd6a6f-093f-45f2-9aeb-d4da4f1ed6b2.png)

I just saw the Yumebooru multiplayer menu and it's closer to in-game menus as in my suggestion.
![image](https://user-images.githubusercontent.com/45118493/157804056-6410191e-46a8-4004-91a8-6d6cac84232e.png)

Ghabry commented:
I actually thought about the opposite: Moving the options from the game browser in a single top row to win more horizontal space for text (game names in the list) 😅.

fdelapena commented:
Aligning values to the right as in Yumebooru modification keeps the content cleaner and more readable for both keys and values. The selectable line cursor helps reading the key-value association when the separation is significant.

Ghabry commented:
I decided to not implement most of the planned functionality by fmatthew because this is too much work to do.

My remaining todo list:

- [ ] Proper storage location for all the stuff (emscripten needs a new mount here to ensure it is remembered on the entire domain)
- [ ] Numeric inputs should use the number input window
- [ ] Frame limiter support
- [ ] Vsync ON/OFF (Restart required)
- [ ] Audio: Global BGM & SE Volume
- [ ] Input: Swapping Primary & Secondary Analog Stick (for Lefties)
- [ ] Input: Swapping AX with BY (for Nintendo fans)

TrunXGIT commented:
Regarding the Video Options:
In Windowed Mode it would be great if you could choose between all integer scalings (x1,x2,x3,x4,x5...) up to the biggest possible that fits inside the display resolution.
In Fullscreen it might be enough to give the option between the biggest possible integer scaling and one that fills the (vertical) screen space completely even if that means uneven/blurred pixels.
The icing on the cake would be an option for shader effects like CRT/LCD ones. But might be out of scope.

aleck099 commented:
@Ghabry How to modify global audio volume?

Mimigris commented:
> Ghabry How to modify global audio volume?

There does not seems to have currently a global audio volume, just BGM or SFX, so you have to manually edit both.

> This is a WIP to add a settings menu. Early feedback is welcome on the general layout and feature set this should have.

Was thinking, maybe adding an option to choose a translation for Easy RPG Menu and settings + translation to load if it exist for a game could be added to the menu ? (and if this is something that can't be edited while in-game exceptions for options could be added)

aleck099 commented:
@Mimigris I saw when playing bgm there is a volume parameter, but how can I modify its volume when the bgm is playing?

Mimigris commented:
> @Mimigris I saw when playing bgm there is a volume parameter, but how can I modify its volume when the bgm is playing?

Not sure if you are really talking here about the setting implementation planned for Easy RPG Player or just talking in terms of coding for a RPG Maker game, so:
Easy RPG Player: Not currently available directly as a build, but this should be available in the settings menu of the Player.
RPG Maker: You have to manually edit each song that play to edit the volume.

If all you want is making your game being less loud when playing you can still change the volume setting of the app in the Volume Mixer of your OS I guess.

Ghabry commented:
Well, imo this is "feature complete" now on Linux if you want to test.

Have to unbreak the build now for the other platforms and provide a config directory everywhere.

Ghabry commented:
Suggested input UI by Jetrotal:

![unknown](https://user-images.githubusercontent.com/1331889/196057933-7eecd806-e3a6-4e46-9099-f52bce2279a7.png)

jetrotal commented:
That suggestions requires always having a system2 hand-icon png.

-----------------------

I'd also suggest limitting the inputs quantity or testing single page per command, 
for cases when the inputs are all huge text chunks like the `JOY_STICK_PRIMARY_UP`

![image](https://user-images.githubusercontent.com/45118493/196058367-d44c8f4e-8d3c-4d5d-a6b1-0df90a622948.png)

Ghabry commented:
Not as colorful as the suggested one but at least fits on one screen ;)

![screenshot_13](https://user-images.githubusercontent.com/1331889/205525537-f666e408-91d0-4da2-a252-4675ecac97ba.png)

-------------

Also pretty close to finishing this (finally!). Only missing is adopting certain homebrew port features to the config scene (stretch, touch ui).

Mimigris commented:
Just found an issue while testing this too:
Opening the settings menu (F1/assigned key - Open Video Options command) while the screen should be hidden will make the screen be shown to the player once said menu has been closed, while this should not be the case. Seems to be a leftover from opening the default menu.

Ghabry commented:
@aleck099

In this PR is a new API on the Audio interface ``BGM_SetGlobalVolume``. This writes in the configuration so it is preserved across sessions.

Right now there is no supported way to do it.

Ghabry commented:
(Not commited yet)

Added a backlight control setting for the 3DS. Yes. There is already one on the touchscreen. They are kept in sync and are saved in the configuration file.

Not tested yet as citra does not emulate this (???).

![a](https://user-images.githubusercontent.com/1331889/211785100-0a223cb3-0af3-412a-a778-d84ee5c04512.png)

Custom names for the keys are now implemented:

![b](https://user-images.githubusercontent.com/1331889/211785105-28901575-ff8e-441b-b5d4-bf9fa7d184e1.png)

![c](https://user-images.githubusercontent.com/1331889/211785106-015a706b-dfa5-45f2-a415-ab0149b52f45.png)

Ghabry commented:
@Mimigris 

Thanks. Screen fade-in is resolved. Missed that edge-case :)

------------

Added the emergency reset sequence (I wanted this option anyway. As I want to prevent the typical RetroArch situation where the Ui becomes unusable due to a button mapping error). It is activated by holding 4 arbitrary keys at once. Is displayed in the input remapping as a hint.

There are still ways to make the Player uncontrollable (by leaving the settings while the input is broken) but this is already close to idiotic user error and you can fix it by (somehow) reentering the settings and doing the reset sequence.

Ghabry commented:
Jenkins: Test this please

Ghabry commented:
First test results:

-  if you swap [AB], it'll swap them then quit the menu (probably because it then consider the "new" input right after), not a major thing but a funny quirk
- The window size is not preserved right now (maybe add this?)
- vsync not saved (?)
- If you put garbage on config.ini none of the keys works
- submenu for key mapping should be a good idea: game, developer/test, other

Ghabry commented:
Suggested by @jetrotal . This was simple to add

![screenshot_3](https://user-images.githubusercontent.com/1331889/212435974-b6a227cf-2e5a-4583-a801-f22b2c36ae6a.png)

aleck099 commented:
@Ghabry That's great

Ghabry commented:
Geting close. Just few bug fixes and platform tests missing.

Getting close.
Redesigned the Game Browser to ensure all menu items fit. You will get used to it.

![screenshot_1](https://user-images.githubusercontent.com/1331889/213334746-068538bb-0996-492a-8e7e-2da52e503253.png)

New settings category "Engine" that contains a few entries that fit nowhere else:

![screenshot_2](https://user-images.githubusercontent.com/1331889/213334813-13d61401-2fe1-41e7-b9e8-847817da26ae.png)

Ghabry commented:
Except for checking some ports this is now feature complete and ready for review. :tada: This took over a year to finish with multiple long breaks because this was so frustrating to program.

Sorry for so many LoC changes and 50+ commits. There was no existing code in the Player to handle persistent configurations so everything had to be written from scratch.

At least further additions of config options will be much less code...

Ghabry commented:
Had to retest Vita and Switch to solve some final bugs.

Only pending tests are Android and a retest on Windows

Mimigris commented:
> Only pending tests are Android and a retest on Windows

The only issue I've found on my side while re-testing the latest version on my Windows is that the cursor while in the Settings menu part where you can remap your keys "Input -> Key/Button mapping" and then _Game / Engine / Developer_ is not properly tracked like the other menus (If your cursor is on the 22th item in the Engine part, when opening the Developer part your cursor will also be on the 22th slot although it is not possible normally).
![screenshot_1](https://user-images.githubusercontent.com/108797333/214651165-5e96884c-f6d7-46d4-9c00-2646562d3215.png)
![screenshot_0](https://user-images.githubusercontent.com/108797333/214651156-594accf6-417d-40ff-a8cd-aaedeb5fc96c.png)

Ghabry commented:
Also finished Android now.

@Mimigris your reported bug should be resolved

fdelapena commented on src/platform/android/org_easyrpg_player_player_EasyRpgPlayerActivity.cpp:
```diff
@@ -2,41 +2,37 @@
  * This file is part of EasyRPG Player
  *
  * Copyright (c) 2017 EasyRPG Project. All rights reserved.
```
At least this header still has the © year, a good chance to remove it

Ghabry replied:
and not touched since 6 years xD
